File: prediction-server.js
/**
 * ============================================================================
 * 🛰️ DRONE DETECTION REAL-TIME PREDICTION SERVER
 * ============================================================================
 * This server handles:
 * 1. 📡 Polling raw hex data from Firebase Realtime Database
 * 2. 🧠 Running ML Inference via ONNX (Drone Detector Model)
 * 3. 🔌 Broadcasting live predictions to Dashboard via Socket.io
 * ============================================================================
 */

require('dotenv').config({ path: '.env.local' });
const ort = require('onnxruntime-node');
const { initializeApp } = require('firebase/app');
const { getDatabase, ref, get } = require('firebase/database');
const http = require('http');
const { Server } = require('socket.io');

// ----------------------------------------------------------------------------
// [1] CONFIGURATION & INITIALIZATION
// ----------------------------------------------------------------------------
const PORT = process.env.PORT || 4000;
const POLLING_INTERVAL = 2000; // 2 Seconds
const MODEL_PATH = './drone_detector_model.onnx';

const firebaseConfig = {
  apiKey: process.env.NEXT_PUBLIC_FIREBASE_API_KEY,
  authDomain: process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN,
  databaseURL: process.env.NEXT_PUBLIC_FIREBASE_DATABASE_URL,
  projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
  storageBucket: process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.NEXT_PUBLIC_FIREBASE_APP_ID,
};

const app = initializeApp(firebaseConfig);
const db = getDatabase(app);
const server = http.createServer();
const io = new Server(server, {
    cors: { origin: "*" } // Allow Dashboard to connect
});

// State Store (In-Memory)
const lastHexMap = {};
const latestPredictionsStore = {};

// ----------------------------------------------------------------------------
// [2] MACHINE LEARNING INFERENCE ENGINE (ML Backend)
// ----------------------------------------------------------------------------
let ortSession;

async function initModel() {
    try {
        console.log('🧠 Loading ONNX Model...');
        ortSession = await ort.InferenceSession.create(MODEL_PATH);
        console.log('✅ Model Loaded Successfully');
    } catch (e) {
        console.error('❌ Failed to load model:', e);
        process.exit(1);
    }
}

/**
 * Perform inference on raw hex data
 * @param {string} hexStr Raw data from sensor
 * @returns {Promise<{id: number, label: string}>}
 */
async function performInference(hexStr) {
    try {
        // Preprocessing: Hex to Integer Array
        let newData = [];
        for (let i = 0; i < hexStr.length; i += 2) {
            newData.push(parseInt(hexStr.substr(i, 2), 16));
        }

        // Feature Engineering: Padding/Trimming to 63 features
        const expectedFeatures = 63; 
        if (newData.length < expectedFeatures) {
            newData = newData.concat(new Array(expectedFeatures - newData.length).fill(0));
        } else if (newData.length > expectedFeatures) {
            newData = newData.slice(0, expectedFeatures);
        }

        // Tensor Creation
        const floatData = new Float32Array(newData);
        const tensor = new ort.Tensor('float32', floatData, [1, expectedFeatures]);

        // Run ML Model
        const results = await ortSession.run({ float_input: tensor }, ['output_label']);
        const prediction = Number(results.output_label.data[0]);

        // Label Mapping
        const mapping = {
            0: { id: 1, label: "WIFI ONLY" },
            1: { id: 2, label: "WIFI+BT" },
            2: { id: 3, label: "DRONE DETECTED" }
        };
        
        return mapping[prediction] || { id: 0, label: "Unknown" };
    } catch (e) {
        console.error('Inference Error:', e);
        return { id: -1, label: "Processing Error" };
    }
}

// ----------------------------------------------------------------------------
// [3] SOCKET & FIREBASE COMMUNICATION (Communication Backend)
// ----------------------------------------------------------------------------

io.on('connection', (socket) => {
    console.log(`🔌 Client connected: ${socket.id}`);
    
    // Sync new client with latest known states
    Object.values(latestPredictionsStore).forEach(pred => {
        socket.emit('prediction_update', pred);
    });

    socket.on('disconnect', () => console.log(`🔌 Client disconnected: ${socket.id}`));
});

async function runPollingLoop() {
    const detectionRef = ref(db, 'detection_system');
    
    console.log(`📡 Polling Started (Interval: ${POLLING_INTERVAL}ms)`);

    setInterval(async () => {
        try {
            const snapshot = await get(detectionRef);
            const nodes = snapshot.val();
            if (!nodes) return;

            for (const [nodeKey, nodeData] of Object.entries(nodes)) {
                let hex = String(nodeData.data_hex || "");
                
                // Hapus prefix "RAW_" agar model ML hanya membaca angkanya
                if (hex.startsWith("RAW_")) {
                    hex = hex.replace("RAW_", "");
                }

                if (!hex) continue;

                // Check if data is new
                if (lastHexMap[nodeKey] !== hex) {
                    lastHexMap[nodeKey] = hex;

                    // Execute Prediction
                    const prediction = await performInference(hex);
                    
                    // Create Payload
                    const payload = {
                        node: nodeKey,
                        prediction_id: prediction.id,
                        prediction_label: prediction.label,
                        timestamp: new Date().toISOString(),
                        original_data: nodeData
                    };

                    // Update Global Store
                    latestPredictionsStore[nodeKey] = payload;

                    // Broadcast to UI
                    io.emit('prediction_update', payload);

                    console.log(`🎯 [${nodeKey}] Prediction: ${prediction.label} | Hex: ${hex.substring(0, 20)}...`);
                }
            }
        } catch (e) {
            console.error('Polling Error:', e);
        }
    }, POLLING_INTERVAL);
}

// ----------------------------------------------------------------------------
// [4] START SERVER
// ----------------------------------------------------------------------------
(async () => {
    await initModel();
    server.listen(PORT, () => {
        console.log(`🚀 Prediction Server running on http://localhost:${PORT}`);
        runPollingLoop();
    });
})();

File: prediction_server.py
import os
import time
import socketio
import eventlet
import requests
from dotenv import load_dotenv
from datetime import datetime

# Impor logika dari TimeseriesPredict
from TimeseriesPredict import predict_timeseries, clean_data_hex

# ----------------------------------------------------------------------------
# [1] CONFIGURATION & INITIALIZATION
# ----------------------------------------------------------------------------
load_dotenv('.env.local')

PORT = int(os.getenv('PORT', 4000))
POLLING_INTERVAL = 2.0  # 2 Seconds
WINDOW_SIZE = 10

# Initialize Socket.IO server
sio = socketio.Server(cors_allowed_origins='*')
app = socketio.WSGIApp(sio)

# State Store (In-Memory)
last_seen_hex = {}
node_buffers = {}
latest_predictions_store = {}

FIREBASE_URL = os.getenv('NEXT_PUBLIC_FIREBASE_DATABASE_URL')
if not FIREBASE_URL:
    print("❌ NEXT_PUBLIC_FIREBASE_DATABASE_URL not found in .env.local")
    exit(1)

def get_firebase_data():
    try:
        url = f"{FIREBASE_URL}/detection_system.json"
        response = requests.get(url)
        if response.status_code == 200:
            return response.json()
    except Exception as e:
        print(f"Firebase fetch error: {e}")
    return None

def map_prediction_to_id(label):
    # Mapping based on time series phase_label
    label = label.lower()
    if "no wifi no drone" in label:
        return 0
    elif "drone" in label:
        return 3
    elif "wifi" in label:
        return 1
    return 0

# ----------------------------------------------------------------------------
# [2] SOCKET COMMUNICATION
# ----------------------------------------------------------------------------
@sio.event
def connect(sid, environ):
    print(f"🔌 Client connected: {sid}")
    # Sync new client with latest known states
    for node, pred in latest_predictions_store.items():
        sio.emit('prediction_update', pred, to=sid)

@sio.event
def disconnect(sid):
    print(f"🔌 Client disconnected: {sid}")

# ----------------------------------------------------------------------------
# [3] POLLING LOOP & BUFFER MANAGEMENT
# ----------------------------------------------------------------------------
def polling_loop():
    print(f"📡 Polling Started (Interval: {POLLING_INTERVAL}s)")
    while True:
        try:
            nodes = get_firebase_data()
            if not nodes:
                eventlet.sleep(POLLING_INTERVAL)
                continue

            for node_key, node_data in nodes.items():
                raw_hex = node_data.get('data_hex', "")
                
                # Hapus prefix "RAW_"
                if raw_hex.startswith("RAW_"):
                    raw_hex = raw_hex.replace("RAW_", "")
                
                if not raw_hex:
                    continue
                
                # Check if data is new
                if last_seen_hex.get(node_key) != raw_hex:
                    last_seen_hex[node_key] = raw_hex
                    
                    # Manage Buffer
                    if node_key not in node_buffers:
                        node_buffers[node_key] = []
                        
                    node_buffers[node_key].append(raw_hex)
                    
                    # Sliding Window
                    if len(node_buffers[node_key]) > WINDOW_SIZE:
                        node_buffers[node_key].pop(0)
                        
                    # Execute Prediction if buffer is full
                    if len(node_buffers[node_key]) == WINDOW_SIZE:
                        print(f"\n🔄 Running prediction for {node_key}...")
                        result = predict_timeseries(node_buffers[node_key])
                        
                        if result:
                            # Create Payload
                            payload = {
                                "node": node_key,
                                "prediction_id": map_prediction_to_id(result["predicted_label"]),
                                "prediction_label": result["predicted_label"],
                                "timestamp": datetime.utcnow().isoformat() + "Z",
                                "original_data": node_data
                            }
                            
                            # Update Global Store
                            latest_predictions_store[node_key] = payload
                            
                            # Broadcast to UI
                            sio.emit('prediction_update', payload)
                            print(f"🎯 [{node_key}] Broadcast: {result['predicted_label']} (Score: {result['cosine_similarity']:.2f})")
                            
        except Exception as e:
            print(f"Polling Error: {e}")
            
        eventlet.sleep(POLLING_INTERVAL)

# ----------------------------------------------------------------------------
# [4] START SERVER
# ----------------------------------------------------------------------------
if __name__ == '__main__':
    print(f"🚀 Prediction Server running on http://localhost:{PORT}")
    
    # Run polling loop in a background thread
    eventlet.spawn(polling_loop)
    
    # Start WSGI server
    eventlet.wsgi.server(eventlet.listen(('', PORT)), app)
